magento/pwa-studio
# V4P2: feat(buildpack): Add Webpack configurator facade

## Description

Pull complex Webpack configuration out of the Venia app and put it in an opinionated Buildpack utility.

This replaces most of the large webpack.config.js file in Venia. The entire PWA Studio runtime depends on the logic in that file; Peregrine, Buildpack, and UpwardJS make assumptions based on that file's configuration, which is an antipattern.

This PR adds a `configureWebpack()` function which returns a Promise for a Webpack configuration object. It takes parameters for a subset of configuration options, including:

- `context`: Absolute path of the project root to ensure consistent behavior no matter the working directory; if webpack.config.js is in project root then this should be simply `__dirname`. 
- `vendor`: Names of modules to bundle into a commons bundle. Set to a sensible default list in Venia.
- `special`: Map of modules to flag as containing 'special' functionality, to be treated differently by the Webpack plugin and loader chain. Keys should be module names. Values should be one or more of the following flags:
    * `esModules`: Denotes that JS code for this module must adhere to the `babel-preset-peregrine` Babel preset. Code will be imported as modules using the same loader chain as local application code, enabling tree-shaking on node_modules dependencies.
    * `cssModules`: Denotes that CSS code for this module must assign all CSS imports to a variable, to receive an object of dynamically generated CSS classes. Style will be imported using the same loader chain as local application stylesheets, enabling total CSS modularity.
    * `graphQLQueries`: Denotes that JS code for this module contains GraphQL queries embedded in calls to the `graphql-tag` template literal. GraphQL will be preprocessed using the same loader chain as local application code.
    * `rootComponents`: Denotes that this module includes a `src/RootComponents` or `lib/RootComponents` directory containing RootComponents to be processed by the `RootComponentsPlugin`. These RootComponents will be added to the map of all RootComponent mappings available to the application.
    * `upward`: Denotes that this module includes an `upward.yml` file at root declaring some subset of server-side behavior it requires and/or supplies. This file will be _merged_, along with the `upward.yml` for any other modules you configure with this flag. It will merge down to a final `upward.yml` file produced as a build asset.

Also adds a lot of debuggability to Buildpack, and cleans up some of its APIs.

Closes #1502 along with #1497 and #1499.

ℹ️ **Docs and tests TODO.**

## Verification Steps

1. In a fresh repo, run `yarn run build`.
2. Examine `packages/venia-concept/webpack.config.js`.
3. Try adjusting the configuration flags to see what happens.

## Proposed Labels for Change Type/Package




## Checklist:
